source/extensions/filters/http/file_system_buffer/fragment.cc

```diff
@@ -0,0 +1,134 @@
+#include "source/extensions/filters/http/file_system_buffer/fragment.h"
+
+#include "source/common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace FileSystemBuffer {
+
+class FragmentData {
+public:
+  virtual bool isMemory() const { return false; }
+  virtual bool isStorage() const { return false; }
+  virtual ~FragmentData() = default;
+};
+
+class MemoryFragment : public FragmentData {
+public:
+  explicit MemoryFragment(Buffer::Instance& buffer);
+  explicit MemoryFragment(Buffer::Instance& buffer, size_t size);
+  std::unique_ptr<Buffer::Instance> extract();
+  bool isMemory() const override { return true; }
+
+private:
+  std::unique_ptr<Buffer::OwnedImpl> buffer_;
+};
+
+class WritingFragment : public FragmentData {};
+
+class ReadingFragment : public FragmentData {};
+
+class StorageFragment : public FragmentData {
+public:
+  explicit StorageFragment(off_t offset) : offset_(offset) {}
+  off_t offset() const { return offset_; }
+  bool isStorage() const override { return true; }
+
+private:
+  const off_t offset_;
+};
+
+Fragment::Fragment(Buffer::Instance& buffer)
+    : size_(buffer.length()), data_(std::make_unique<MemoryFragment>(buffer)) {}
+
+Fragment::Fragment(Buffer::Instance& buffer, size_t size)
+    : size_(size), data_(std::make_unique<MemoryFragment>(buffer, size)) {}
+
+Fragment::~Fragment() = default;
+bool Fragment::isMemory() const { return data_->isMemory(); }
+bool Fragment::isStorage() const { return data_->isStorage(); }
+
+MemoryFragment::MemoryFragment(Buffer::Instance& buffer)
+    : buffer_(std::make_unique<Buffer::OwnedImpl>()) {
+  buffer_->move(buffer);
+}
+
+MemoryFragment::MemoryFragment(Buffer::Instance& buffer, size_t size)
+    : buffer_(std::make_unique<Buffer::OwnedImpl>()) {
+  buffer_->move(buffer, size);
+}
+
+std::unique_ptr<Buffer::Instance> Fragment::extract() {
+  ASSERT(isMemory());
+  auto ret = dynamic_cast<MemoryFragment*>(data_.get())->extract();
+  data_.reset();
+  size_ = 0;
+  return ret;
+}
+
+std::unique_ptr<Buffer::Instance> MemoryFragment::extract() { return std::move(buffer_); }
+
+absl::StatusOr<CancelFunction>
+Fragment::toStorage(AsyncFileHandle file, off_t offset,
+                    std::function<void(absl::StatusOr<UpdateFragmentFunction>)> on_done) {
+  ASSERT(isMemory());
+  auto data = dynamic_cast<MemoryFragment*>(data_.get())->extract();
```

Comment:
Instead of using dynamic_cast to deal with the state machine, and potential errors show up as nullptr, can you use absl::Variant instead? Then the state machine and possible values are more clear?

Reply:
Nice suggestion, thanks. Not so keen on more internal implementation details being moved into the header file, which unfortunately is required for variant, but it's better enough to be worth it.